src/Simulator/IncrementalRedstoneSimulator/RedstoneHandler.h

```diff
@@ -0,0 +1,127 @@
+
+#pragma once
+
+#include "World.h"
+#include "Vector3.h"
+
+
+
+
+
+class cRedstoneHandler
+{
+public:
+
+	cRedstoneHandler(cWorld & a_World) :
+		m_World(a_World)
+	{
+	}
+
+public:
+
+	// Disable the copy constructor and assignment operator
+	DISALLOW_COPY_AND_ASSIGN(cRedstoneHandler);
+
+	struct PoweringData
+	{
+	public:
+		PoweringData(BLOCKTYPE a_PoweringBlock, unsigned char a_PowerLevel) :
+			PoweringBlock(a_PoweringBlock),
+			PowerLevel(a_PowerLevel)
+		{
+		}
+
+		PoweringData(void) :
+			PoweringBlock(E_BLOCK_AIR),
+			PowerLevel(0)
+		{
+		}
+
+		BLOCKTYPE PoweringBlock;
+		unsigned char PowerLevel;
+
+		inline friend bool operator < (const PoweringData & a_Lhs, const PoweringData & a_Rhs)
+		{
+			return (
+				(a_Lhs.PowerLevel < a_Rhs.PowerLevel) ||
+				(
+					(a_Lhs.PowerLevel == a_Rhs.PowerLevel) &&
+					((a_Lhs.PoweringBlock == E_BLOCK_REDSTONE_WIRE) && (a_Rhs.PoweringBlock != E_BLOCK_REDSTONE_WIRE))
+				)
+			);
+		}
+
+		inline friend bool operator == (const PoweringData & a_Lhs, const PoweringData & a_Rhs)
+		{
+			return (a_Lhs.PowerLevel == a_Rhs.PowerLevel);
+		}
+
+		inline friend bool operator != (const PoweringData & a_Lhs, const PoweringData & a_Rhs)
+		{
+			return !operator ==(a_Lhs, a_Rhs);
+		}
+	};
+
+	virtual cVector3iArray Update(const Vector3i & a_Position, BLOCKTYPE a_BlockType, NIBBLETYPE a_Meta, PoweringData a_PoweringData) = 0;
+	virtual unsigned char GetPowerDeliveredToPosition(const Vector3i & a_Position, BLOCKTYPE a_BlockType, NIBBLETYPE a_Meta, const Vector3i & a_QueryPosition, BLOCKTYPE a_QueryBlockType) = 0;
+	virtual unsigned char GetPowerLevel(const Vector3i & a_Position, BLOCKTYPE a_BlockType, NIBBLETYPE a_Meta) = 0;
+	virtual cVector3iArray GetValidSourcePositions(const Vector3i & a_Position, BLOCKTYPE a_BlockType, NIBBLETYPE a_Meta) = 0;
+
+	// Force a virtual destructor in descendants
+	virtual ~cRedstoneHandler() {}
+
+protected:
+
+	cWorld & m_World;
+
+	template <class Container>
+	Container StaticAppend(const Container & a_Lhs, const Container & a_Rhs)
```

Comment:
This function would be far better written with rvalue references:
```
template <class Container>
Container StaticAppend(Container && a_Lhs, Container && a_Rhs)
{
	a_Lhs.insert(a_Lhs.end(), a_Rhs.begin(), a_Rhs.end());
	return a_Lhs;
}
```
This way, if a_Lhs has the capacity, or is a node based container, you don't end up copying the elements of a_Rhs.